falcon/cors.py

```diff
@@ -0,0 +1,426 @@
+            self._set_allow_origin(resp, origin)
+            return True
+
+        regex = self._cors_config['allow_origins_regex']
+        if regex is not None:
+            if regex.match(origin):
+                self._set_allow_origin(resp, origin)
+                return True
+
+        return False
+
+    def _process_allow_headers(self, req, resp, requested_headers):
+        """Adds the Access-Control-Allow-Headers header to the response,
+        using the cors settings to determine which headers are allowed.
+
+        Returns:
+            True if all the headers the client requested are allowed.
+            False if some or none of the headers the client requested are allowed.
+        """
+        if not requested_headers:
+            return True
+        elif self._cors_config['allow_all_headers']:
+            self._set_allowed_headers(resp, requested_headers)
+            return True
+
+        approved_headers = []
+        for header in requested_headers:
+            if header in self._cors_config['allow_headers_list']:
+                approved_headers.append(header)
+            elif self._cors_config.get('allow_headers_regex'):
+                if self._cors_config['allow_headers_regex'].match(header):
+                    approved_headers.append(header)
+
+        if len(approved_headers) == len(requested_headers):
+            self._set_allowed_headers(resp, approved_headers)
+            return True
+
+        return False
+
+    def _process_methods(self, req, resp, resource):
+        """Adds the Access-Control-Allow-Methods header to the response,
+        using the cors settings to determine which methods are allowed.
+        """
+        requested_method = self._get_requested_method(req)
+        if not requested_method:
+            return False
+
+        if self._cors_config['allow_all_methods']:
+            allowed_methods = self._get_resource_methods(resource)
+            self._set_allowed_methods(resp, allowed_methods)
+            if requested_method in allowed_methods:
+                return True
+        elif requested_method in self._cors_config['allow_methods_list']:
+            resource_methods = self._get_resource_methods(resource)
+            # Only list methods as allowed if they exist
+            # on the resource AND are in the allowed_methods_list
+            allowed_methods = [
+                method for method in resource_methods
+                if method in self._cors_config['allow_methods_list']
+            ]
+            self._set_allowed_methods(resp, allowed_methods)
+            if requested_method in allowed_methods:
+                return True
+
+        return False
+
+    def _get_resource_methods(self, resource):
+        allowed_methods = []
+        for method in HTTP_METHODS:
+            if hasattr(resource, 'on_' + method.lower()):
+                allowed_methods.append(method)
+        return allowed_methods
```

Comment:
I'm intentionally not using a list comprehension because I find it hard to read when combined with what's going on in hasattr() and lower().  Also this snippet was inspired by https://github.com/falconry/falcon/blob/master/falcon/routing/util.py#L104